src/core/initialize.rs

```diff
@@ -80,40 +80,73 @@ impl StackTraceGetter {
+                                     }
+                                     #[cfg(target_os = "windows")]
+                                     std::io::ErrorKind::PermissionDenied => {
+                                         return Err(MemoryCopyError::ProcessEnded)
+                                     }
+                                     _ => {}
+                                 };
+                             }
+                             #[cfg(target_os = "linux")]
+                             Ok(remoteprocess::Error::NixError(e)) => match e {
+                                 nix::Error::Sys(nix::errno::Errno::EPERM) => {
+                                     return Err(MemoryCopyError::ProcessEnded);
+                                 }
+                                 nix::Error::Sys(nix::errno::Errno::ESRCH) => {
+                                     match Process::new(self.process.pid) {
+                                         Ok(_) => {
+                                             debug!("Process {} is actually alive, spurious ESRCH detected", self.process.pid);
+                                             return Err(MemoryCopyError::ProcessNotLocked);
+                                         },
+                                         Err(e) => {
+                                             debug!("ESRCH Process {} is not alive: {:?}", self.process.pid, e);
+                                             return Err(MemoryCopyError::ProcessEnded);
+                                         }
+                                     }
+                                 }
+                                 _ => {}
+                             },
+                             _ => {}
+                         }
+                     }
+                     Err(e.into())
+                 })
+        };
+
         let _lock;
         if self.lock_process {
-            _lock = self
-                .process
-                .lock()
-                .context("locking process during stack trace retrieval")
-                .or_else(|e| {
-                    if let Some(source) = e.source() {
-                        match source.downcast_ref::<remoteprocess::Error>().ok_or(source) {
-                            Ok(remoteprocess::Error::IOError(e)) => {
-                                match e.kind() {
-                                    std::io::ErrorKind::NotFound => {
-                                        return Err(MemoryCopyError::ProcessEnded)
-                                    }
-                                    // Windows
-                                    std::io::ErrorKind::PermissionDenied => {
-                                        return Err(MemoryCopyError::ProcessEnded)
-                                    }
-                                    _ => {}
-                                };
+            let mut retries = 10;
```

Comment:
retries should be very uncommon, as they would necessitate the lock failing on ESRCH repeatedly, so each successive occurrence of this should be less and less probable. In reality, one retry is probably enough most of the time, and only in exceptional circumstances should the 10 retries here be hit

Reply:
From the test case that @acj wrote, I found I needed to bump this to 20 retries to pass the test. I don't expect it is likely to see such pathological cases in the wild, but with 20 we should be confident that the retries will be sufficient.